src/Tuple.js

```diff
@@ -0,0 +1,74 @@
+function Tuple(x, y) {
+    switch (arguments.length) {
+        case 0:
+            throw new TypeError('no arguments to Tuple');
+        case 1:
+            return function(y) {
+                return new _Tuple(x, y)
+            };
+        default:
+            return new _Tuple(x,y)
+    }
+};
+
+function _Tuple(x, y) {
+    this[0] = x;
+    this[1] = y;
+    this.length = 2;
+};
+
+function ensureConcat() {
+    var xs = [].slice.call(arguments)
+    xs.map(function(x){
+      if(typeof x.concat != "function") {
+          var display = x.show ? x.show() : x;
+          throw new TypeError(display + " must be a semigroup to perform this operation");
+      };
+    })
```

Comment:
this `map` for side-effects feels a bit strange, but i wonder if making `ensureConcat` return a Bool if it is concatable is any better. I don't think so...